{% import 'utils.html' as utils %}

{% extends 'base.html' %}

{% block header %}
<script src="{{ url_for('static', filename='pileup.js') }}"></script>
<link rel="stylesheet" href="{{ url_for('static', filename='pileup.css') }}" ></link>
<script type='text/javascript' src='{{ url_for("static", filename="bio-pv.min.js") }} '></script>

{% endblock %}

{% block content %}

{% if run['results']=={} %}
	<div class="alert alert-warning mt-4" role="alert">
        Your run ID is: <b>{{run["_id"]}}</b>. This run is still processing, this usually takes a 
        few seconds for a fasta file. For an approximate runtime using fastq data please compare 
        the number of sequences with the plot below.</div>
    <div class="text-center">
       <img src="{{url_for('static', filename='mapping_time.png')}}" alt="" width="50%">
    </div>
    <script>
        setTimeout(function(){
            window.location.reload(1);
        }, 5000);
    </script>
{% else %}

<div class="row justify-content-md-center">
    <div class="col-md-10">
        <div class="card border-dark mb-3">
    		<div class="card-header">Summary</div>
    		<div class="card-body text-dark">
    			<p class="card-text"><b>Run ID:</b> {{ run["_id"] }}</p>
                <p class="card-text"><b>Data type:</b> {{ run["data_type"] }}</p>
    			<p class="card-text"><b>Sample name:</b> {{ run["sample_name"] }}</p>
    			<p class="card-text"><b>Date:</b> {{ run["created"] }}</p>
    			<p class="card-text"><b>Strain:</b> {{ run["results"]["clade"] }}</p>
                <p class="card-text"><b>Number of variants:</b> {{ run["results"]["num_variants"] }}</p>
                {% if run["data_type"]=="fastq" %}
                    <p class="card-text"><b>Mean depth:</b> {{ run["results"]["mean_depth"] }}</p>
                {% endif %}

    			<hr>
                <p>
                    Download files: <a href="{{ url_for('static', filename='results/') + run['_id']}}.zip"><img src="{{url_for('static', filename='zip-icon.svg')}}" alt="" width="45" height="52"></a>
                </p>

    		</div>
    	</div>
    </div>

</div>

<div class="row justify-content-md-center">
    <div class="col-md-10">
        <div class="card border-dark mb-3">
            <div class="card-header">Mutations</div>
            {% if run["results"]["num_variants"]==0 %}
                <div class="row">
                    <div class="col-md-12 text-center">
                        No mutations found for this sample.

                    </div>
                </div>
            {% else %}
                <table class="table table-hover" id="results_table">
                    <thead>
                        <tr>
                            <!-- <th scope="col"></th> -->
                            <th scope="col">Genomic position</th>
                            <th scope="col">Gene</th>
                            <th scope="col">Type</th>
                            <th scope="col">Coding change</th>
                        </tr>
                    </thead>

                    <tbody id="resultsTable">
                        {% for mutation in run["results"]["variants"] %}
                            <tr name="{{ mutation['pos'] }}">
                                <td><a href="{{ url_for('results.mutation', position=mutation['pos']) }}">{{ mutation['pos'] }}</a></td>
                                <td>{{ mutation['gene'] }}</td>
                                <td>{{ mutation['types'] }}</td>
                                <td>{{ mutation['changes'] }}</td>
                            </tr>
                        {% endfor %}
                    </tbody>
                </table>
            {% endif %}
        </div>
    </div>
</div>

{% if run["data_type"]=="fastq" and run["results"]["mean_depth"]<100 %}
<div class="row justify-content-md-center">
    <div class="col-md-10">
        <div class="card border-dark mb-3">
    		<div class="card-header">Pileup</div>
    		<div class="card-body text-dark">
    			<div id="pileup"></div>
    		</div>
    	</div>
    </div>
</div>
{% endif %}


<div class="row justify-content-md-center" >
    <div class="col-md-10">
        <div class="card border-dark mb-3">
    		<div class="card-header">Protein Structure</div>
    		<div class="card-body text-dark">
                <div class="">
                    Mutations present proteins and regions that have been structurally
                    characterised can be viewed with there representative structues here.
                    The chain on which the mutation is placed is highlighted in blue.
                </div>
                <hr>
                {% if structure_variants|length > 0 %}
                    <div class="row">
                        <div class="col-md-12" id="structure-mutations">
                            {% for mutation in structure_variants %}
                                <button type="button" class="btn btn-outline-primary" onclick="selectAndColourStructure(this)">{{ mutation["gene"] }} {{ mutation["changes"] }}</button>
                            {% endfor %}
                        </div>
                    </div>
                    <hr>
                    <div id="structure-viewer" style="height: 500px; overflow:hidden;"></div>
                {% else %}
                    <div class="">
                        No eligible mutations found.
                    </div>
                {% endif %}


    		</div>
            <div id="structure-footer" class="card-footer text-muted text-center">
                PDB code:
            </div>
    	</div>
    </div>
</div>


<script type="text/javascript">
    json_result = {{ run|tojson|safe }}
    variants = {}
    json_result.results.variants.forEach((item, i) => {
        variants[item.pos] = item
    });
    console.log(variants)


	{% if run["data_type"]=="fastq" and run["results"]["mean_depth"]<100 %}

	    var bamSource = pileup.formats.bam({
	        url: "{{ url_for('static', filename='results/') + run['_id']}}.bam",
	        indexUrl: "{{ url_for('static', filename='results/') + run['_id']}}.bam.bai",
	    });
	    var div = document.getElementById('pileup');
	    var p = pileup.create(div, {
	      range: {contig: 'NC_045512.2', start: 1, stop: 250},
	      tracks: [
	        {
	          viz: pileup.viz.genome(),
	          isReference: true,
	          data: pileup.formats.twoBit({
	            url: "{{ url_for('static', filename='cvdb.2bit')}}"
	          }),
	          name: 'Reference'
	        },
	        {
	            viz: pileup.viz.variants(),
	            data: pileup.formats.vcf({
	              url: "{{ url_for('static', filename='results/') + run['_id']}}.vcf"
	            }),
	            options: {
	              onVariantClicked: function(data) {
	                var content="";
	                for (let [key, value] of Object.entries(variants[Number(data[0].vcfLine.split("\t")[1])])) {
	                    content+= `${key}: ${value}`+"\n";
	                }
	                content+="\n"
	                alert(content);
	              },
	            },
	            name: 'Variants'
	          },
	        {
	          viz: pileup.viz.coverage(),
	          data: bamSource,
	          cssClass: 'normal',
	          name: 'Coverage'
	        },
	        {
	            viz: pileup.viz.pileup(),
	            data: bamSource,
	            cssClass: 'normal',
	            name: 'Alignments'
	        },
	        // ...
	      ]
	    });
	{% endif %}



    var structure_info = {{ structures|tojson|safe }}


    var parent = document.getElementById('structure-viewer')
    var options = {
      width: parent.offsetWidth,
      height: 500,
      antialias: true,
      quality : 'medium'
    };
    // insert the viewer under the Dom element with id 'gl'.
    var viewer
    var geom;
    var a
    function color_residue(chain,num) {
        console.log("Colouring chain "+chain+" residue "+num)
        return new pv.color.ColorOp(function(atom, out, index) {
          // index + 0, index + 1 etc. are the positions in the output array
          // at which the red (+0), green (+1), blue (+2) and  alpha (+3)
          // components are to be written.

            if (atom.residue().chain().name()==chain){
                if (atom.residue().num() == num) {
            // console.log(atom)
                    out[index+0] = 1.0; out[index+1] = 0.0;
                    out[index+2] = 0.0; out[index+3] = 1.0;
                } else {
                    out[index+0] = 0.6; out[index+1] = 0.6;
                    out[index+2] = 1; out[index+3] = 1.0;
                }
          } else {
            out[index+0] = 1; out[index+1] = 1;
            out[index+2] = 1; out[index+3] = 1.0;
          }
        });
    }

    function loadStructure(pdb_file, chain,residue_num) {
        parent.innerHTML = "";
        viewer = pv.Viewer(parent, options);

        console.log("Loading structure "+pdb_file)
      // asynchronously load the PDB file for the dengue methyl transferase
      // from the server and display it in the viewer.

      pv.io.fetchPdb(pdb_file, function(structure) {
          // display the protein as cartoon, coloring the secondary structure
          // elements in a rainbow gradient.
          geom = viewer.cartoon('protein', structure, { color : color.uniform([0.9,0.9,0.9,1]) });

          // there are two ligands in the structure, the co-factor S-adenosyl
          // homocysteine and the inhibitor ribavirin-5' triphosphate. They have
          // the three-letter codes SAH and RVP, respectively. Let's display them
          // with balls and sticks.
          viewer.autoZoom();
          highlight_residue(chain,residue_num)
      });



    }

    function highlight_residue(chain,num){
        geom.colorBy(color_residue(chain,num))
        viewer.requestRedraw()
    }


    function selectAndColourStructure(objButton){
        tmp = objButton.innerHTML.split(" ")
        gene = tmp[0]
        mutation = tmp[1]
        pos = Number(mutation.match(/\d+/)[0])
        if (structure_info.hasOwnProperty(gene)){
            document.getElementById('structure-footer').innerHTML = `<b>Mutation:</b> ${gene} ${mutation} - <b>PDB code:</b> <a href='https://www.rcsb.org/structure/${structure_info[gene]["pdb_code"]}'>${structure_info[gene]["pdb_code"]}</a>  <b>Chain:</b> ${structure_info[gene]["chain"]}`
            loadStructure(structure_info[gene]["pdb_file"],structure_info[gene]["chain"],pos)
        }

    }

    // load the methyl transferase once the DOM has finished loading. That's
    // the earliest point the WebGL context is available.
    {% if structure_variants|length > 0 %}
        document.addEventListener('DOMContentLoaded', function(){
            document.getElementById("structure-mutations").childNodes[1].click()
        });
    {% endif %}




</script>


{% endif %}

{% endblock %}
